### PR TITLE
New Quickstart template for AKS cluster with a private endpoint for Kubernetes API server and Azure Firewall integration for all ingress & egress traffic

### PR DESCRIPTION
**Pull request by cbellee**
# Best Practice Checklist

Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment.
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by cbellee** [comment]
Hi, It seems that when trying to link the DNS zone automatically created in the AKS managed 'MC_*' resource group to a vNet in the customer managed resource group, the test deployment fails due to lack of permissions. Although this works successfully in my own subscription.

_"Message=The template deployment failed with error: 'Authorization failed for template resource 'linkedTemplate-privateDnsLink' of type 'Microsoft.Resources/deployments'. The client '5991c0de-1362-4164-8ac0-0ab3b91175ee' with object id '5991c0de-1362-4164-8ac0-0ab3b91175ee' does not have permission to perform action 'Microsoft.Resources/deployments/write' at scope '/subscriptions/0cec7090-2e08-4498-9337-eb96ade50821/resourceGroups/MC_azdo-11754029-143a-41d9-98bf-6acc99dc3751_aks-f69d_westus/providers/Microsoft.Resources/deployments/linkedTemplate-privateDnsLink'."_

@bmoore-msft

**Comment by bmoore-msft** [comment]
What permissions are needed to deploy the sample?  That looks like a RG that's created by one of the resources (AKS?) which would imply subscription owner?

**Comment by cbellee** [comment]
I believe you'd need a minimum of Contributor RBAC permissions on the MC_* resource group to create a new deployment.

**Comment by bmoore-msft** [comment]
Since that RG is created by another principal during deployment it sounds like the user needs to be a subscription contributor (if not owner).  let's do the following:

* update the readme to indicate the perms the principal deploying the template needs to have (e.g. in the prereqs section)
* add some PowerShell/CLI commands to help users accomplish all the steps in the prereqs
* set the validationType property in metadata.json to "Manual" as we'll have to run manual validation on it

File: 301-aks-private-endpoint-firewall/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+        },
+        {
+            "comments": "This deployment links to the firewall.json nested ARM template",
+            "name": "linkedTemplate-firewall",
+            "type": "Microsoft.Resources/deployments",
+            "apiVersion": "2019-10-01",
+            "dependsOn": [
+                "vnetCopy"
+            ],
+            "resourceGroup": "[resourceGroup().name]",
```
Comment (bmoore-msft) [inline 301-aks-private-endpoint-firewall/azuredeploy.json:246]: remove this property

File: 301-aks-private-endpoint-firewall/azuredeploy.parameters.json [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            "value": "GEN-SSH-PUB-KEY"
+        },
+        "aadAdminGroupObjectIds": {
+            "value": [
+                "GEN-UNIQUE"
+            ]
+        },
+        "adminUsername": {
+            "value": "localadmin"
+        }
```
Comment (bmoore-msft) [inline 301-aks-private-endpoint-firewall/azuredeploy.parameters.json:85]: use GEN-UNIQUE

File: 301-aks-private-endpoint-firewall/azuredeploy.parameters.json [hunk 3]
```diff
@@ -0,0 +1,87 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "vNets": {
+            "value": [
```
Comment (bmoore-msft) [inline 301-aks-private-endpoint-firewall/azuredeploy.parameters.json:6]: remove literal values and use defaultValues in the template

File: 301-aks-private-endpoint-firewall/nestedtemplates/aks.json [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+            "minValue": 1,
+            "maxValue": 50,
+            "type": "int",
+            "metadata": {
+                "description": "The maximum number of agent nodes for the cluster."
+            }
+        },
+        "aksNodeVMSize": {
+            "defaultValue": "Standard_F2",
+            "allowedValues": [
```
Comment (bmoore-msft) [inline 301-aks-private-endpoint-firewall/nestedtemplates/aks.json:71]: remove allowedValues

File: 301-aks-private-endpoint-firewall/nestedtemplates/aks.json [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+                "Standard_NV24",
+                "Standard_NV6"
+            ],
+            "type": "string",
+            "metadata": {
+                "description": "The size of the Virtual Machine."
+            }
+        },
+        "aksVersion": {
+            "defaultValue": "1.16.9",
```
Comment (bmoore-msft) [inline 301-aks-private-endpoint-firewall/nestedtemplates/aks.json:284]: remove defaultValues where you are passing through param values

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Pull request closed without merging**